pkg/agent/controller/networkpolicy/priority.go

```diff
@@ -281,3 +281,44 @@ func (pa *priorityAssigner) Release(ofPriority uint16) {
 	idxToDel := sort.Search(len(pa.sortedOFPriorities), func(i int) bool { return ofPriority <= pa.sortedOFPriorities[i] })
 	pa.sortedOFPriorities = append(pa.sortedOFPriorities[:idxToDel], pa.sortedOFPriorities[idxToDel+1:]...)
 }
+
+// RevertReassignments reverts priority updates previously computed by the priorityAssigner,
+// if the reassignments failed at the OVS level. Note that it assumes the priortiy updates in
+// the input is not empty and consecutive，e.g. { 100: 99, 99: 98, 98: 97 }
+func (pa *priorityAssigner) RevertReassignments(updates map[uint16]uint16) {
+	siftedUp := false
+	originals := make([]uint16, len(updates))
+	idx := 0
+	for original, updated := range updates {
+		// Use any one of the map entries to determine if the original update was from a siftUp or siftDown operation.
+		if idx == 0 && updated > original {
+			siftedUp = true
+		}
+		originals[idx] = original
+		idx++
+	}
+	// If the original operation was siftUp, the priority reversion should start from the lowest priority
+	// updates and work up to the highest priority updates, in order. This is to ensure that when querying
+	// the original ofPriority, the Priority value was not overridden by other reversion operations.
```

Comment:
Could you help explain more about "This is to ensure that when querying the original ofPriority, the Priority value was not overridden by other reversion operations."?
I don't get "querying the original ofPriority" here, and I think the GetOFPriority and RevertReassignments operation is called with lock, so no other reversion operations could happen in-between?

Reply:
This is simply explaining why we need to sort the updates, it is not related to any race conditions. Let's say we have a current map {100: p1, 99: p2, 98: p3} and updates to be reverted {98: 99, 99: 100}. We know that after reversion, 1) p3 should be release because it was the insertion of p3 that led to p2 being pushed to 99 and p1 pushed to 100. 2) after reversion p2 should sit on 98 and p1 should sit on 99. This is what `p := pa.ofPriorityMap[updated]   pa.ofPriorityMap[original] = p` does. 
However, this operation cannot be done with random order. In this particular case, if we revert {99: 100} first, by the end of the reversion we will have {99: p1, 98: p3} and p2 would be 'overriden'. By the time we consider {98: 99} we cannot determine the correct Priority to be put on 98. Hence, the updates needs to be sorted or reversely sorted first, based on whether the original operation was a siftUp or a siftDown, and the reversion needs to be performed in order.

Reply:
Thanks for the explanation! Now I understand how this works, but honestly it takes me quite some time. 
I want to discuss if another approach could make this simpler: could it include all the reversion logic in a closure and make "GetOFPriority" return it along with the assigned priority and the priority reassignments? Then maybe we can get rid of the direction check, the sorting, and it might be clearer to create the reversion func right after making the priority reassignment. what do you think?

Reply:
Thanks for the suggestion. I uploaded a new commit which does just that. The good side is it gets rid of the direction check and sorting, but it also adds another return value (the closure) to a bunch of functions that already has three returns. I think I am okay with both solutions, and we can merge the closure implementation but you think it is a cleaner approach.

Reply:
Thanks for making the change. It looks clearer to me.